apps/runtime/src/main/java/quarks/apps/runtime/MonitorApp.java

```diff
@@ -0,0 +1,204 @@
+ * them.</p>
+ * <p> 
+ * The monitoring application must be submitted within a context which 
+ * provides the following services:
+ * <ul>
+ * <li>ApplicationService - an {@code ApplicationServiceMXBean} control 
+ * registered by this service is used to resubmit failed applications.</li>
+ * <li>ControlService - the application queries this service for an 
+ * {@code ApplicationServiceMXBean} control, which is then used for 
+ * restarting failed applications.</li>
+ * <li>JobRegistryService - generates job monitoring events. </li>
+ * </ul>
+ * </p>
+ */
+public class MonitorApp {
+    private final TopologyProvider provider;
+    private final DirectSubmitter<Topology, Job> submitter;
+    private final Topology topology;
+    private static final Logger logger = LoggerFactory.getLogger(MonitorApp.class);
+
+    /**
+     * Constructs a {@code MonitorApp} with the specified name in the 
+     * context of the specified provider.
+     * 
+     * @param provider the topology provider
+     * @param submitter a {@code DirectSubmitter} which provides required 
+     *      services and submits the application
+     * @param name the application name
+     * 
+     * @throws IllegalArgumentException if the submitter does not provide 
+     *      access to the required services
+     */
+    public MonitorApp(TopologyProvider provider, 
+            DirectSubmitter<Topology, Job> submitter, String name) {
+
+        this.provider = provider;
+        this.submitter = submitter;
+        validateSubmitter();
+        this.topology = declareTopology(name);
+    }
+    
+    /**
+     * Submits the application topology.
+     * 
+     * @return the job.
+     * @throws InterruptedException
+     * @throws ExecutionException
+     */
+    public Job submit() throws InterruptedException, ExecutionException {
+        Future<Job> f = submitter.submit(topology);
+        return f.get();
+    }
+
+    /**
+     * Submits an application using an {@code ApplicationServiceMXBean} control 
+     * registered with the specified {@code ControlService}.
+     * 
+     * @param applicationName the name of the application to submit
+     * @param controlService the control service
+     */
+    public static void submitApplication(String applicationName, ControlService controlService) {
+        try {
+            Set<ApplicationServiceMXBean> controls = 
+                    controlService.getControls(ApplicationServiceMXBean.class);
+            if (controls.isEmpty()) {
+                throw new IllegalStateException(
+                        "Could not find a registered control with the following interface: " + 
+                        ApplicationServiceMXBean.class.getName());                
+            }
+            for (ApplicationServiceMXBean control : controls)
+// TODO add ability to submit with the initial application configuration
+                control.submit(applicationName, null);
```

Comment:
Just trying to understand / imagine when there might be more than one ApplicationServiceMXBean registered and then what it might mean to submit the same applicationName multiple times once for each AS control?